stetho/src/main/java/com/facebook/stetho/inspector/protocol/module/DOM.java

```diff
@@ -2,30 +2,315 @@
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+import javax.annotation.Nullable;
+import javax.annotation.concurrent.GuardedBy;
+
 public class DOM implements ChromeDevtoolsDomain {
-  public DOM() {
+  private final Provider mDOMProvider;
+  private final DOMPeerManager mPeerManager;
+  private final ObjectMapper mObjectMapper;
+
+  private final Object mNodeMapSync = new Object();
+
+  @GuardedBy("mNodeMapSync")
+  private int mNextNodeId = 1;
+
+  @GuardedBy("mNodeMapSync")
+  private final Map<Object, Integer> mElementToNodeIdMap = new HashMap<>();
+
+  @GuardedBy("mNodeMapSync")
+  private final SparseArray<Object> mNodeIdToElementMap = new SparseArray<>();
+
+  public DOM(Provider provider) {
+    mDOMProvider = Util.throwIfNull(provider);
+    mDOMProvider.setListener(new ProviderListener());
+
+    mPeerManager = new DOMPeerManager();
+    mObjectMapper = new ObjectMapper();
+  }
+
+  private static final class DOMPeerManager extends ChromePeerManager {
+    public DOMPeerManager() {
+    }
+  }
+
+  private final class ProviderListener implements Provider.Listener {
+    @Override
+    public void onAttributeModified(Object element, String name, String value) {
+      AttributeModifiedEvent message = new AttributeModifiedEvent();
+      message.nodeId = getOrCreateNodeIdForElement(element);
+      message.name = name;
+      message.value = value;
+      mPeerManager.sendNotificationToPeers("DOM.attributeModified", message);
+    }
+
+    @Override
+    public void onAttributeRemoved(Object element, String name) {
+      AttributeRemovedEvent message = new AttributeRemovedEvent();
+      message.nodeId = getOrCreateNodeIdForElement(element);
+      message.name = name;
+      mPeerManager.sendNotificationToPeers("DOM.attributeRemoved", message);
+    }
+
+    @Override
+    public void onChildInserted(Object parentElement, Object previousElement, Object childElement) {
+      ChildNodeInsertedEvent message = new ChildNodeInsertedEvent();
+      message.parentNodeId = getOrCreateNodeIdForElement(parentElement);
+
+      message.previousNodeId = (previousElement == null)
+          ? -1  // using -1 was just a guess, and it seemed to work. this is for the case where we
+                // go from 0 to 1 children, or if we just happen to be inserting at index 0
+          : getOrCreateNodeIdForElement(previousElement);
+
+      message.node = createNodeForElement(childElement);
+      mPeerManager.sendNotificationToPeers("DOM.childNodeInserted", message);
+    }
+
+    @Override
+    public void onChildRemoved(Object parentElement, Object childElement) {
+      ChildNodeRemovedEvent message = new ChildNodeRemovedEvent();
+      message.parentNodeId = getOrCreateNodeIdForElement(parentElement);
+      message.nodeId = getOrCreateNodeIdForElement(childElement);
+      mPeerManager.sendNotificationToPeers("DOM.childNodeRemoved", message);
+      unhookElement(childElement);
```

Comment:
I just realized there is a serious bug here in that you do not run any logic in `onLastPeerUnregistered`.  This should be leaking, no?